File: test/language/explicit_parameters/python/ExplicitSameAsFieldTest.py
import unittest
import os

from testutils import getZserioApi, getApiDir

class ExplicitSameAsFieldTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.api = getZserioApi(__file__, "explicit_parameters.zs")
        cls._fileName = os.path.join(getApiDir(os.path.dirname(__file__)), "explicit_same_as_field_test.sqlite")

        test_api = cls.api.explicit_same_as_field
        class SameAsFieldTableParameterProvider(test_api.SameAsFieldTable.IParameterProvider):
            @staticmethod
            def getCount(_row):
                return SAME_AS_FIELD_TABLE_COUNT_EXPLICIT

        cls.SameAsFieldTableParameterProvider = SameAsFieldTableParameterProvider

    def setUp(self):
        if os.path.exists(self._fileName):
            os.remove(self._fileName)
        self._database = self.api.ExplicitParametersDb.fromFile(self._fileName)
        self._database.createSchema()

    def tearDown(self):
        self._database.close()

    def testDeleteTable(self):
        self.assertTrue(self._isTableInDb())

        testTable = self._database.getSameAsFieldTable()
        testTable.deleteTable()
        self.assertFalse(self._isTableInDb())

        testTable.createTable()
        self.assertTrue(self._isTableInDb())

    def testReadWithoutCondition(self):
        testTable = self._database.getSameAsFieldTable()

        writtenRows = self._createSameAsFieldTableRows()
        testTable.write(writtenRows)

        parameterProvider = self.SameAsFieldTableParameterProvider()
        readRows = testTable.read(parameterProvider)
        numReadRows = 0
        for readRow in readRows:
            self._checkSameAsFieldTableRow(writtenRows[numReadRows], readRow)
            numReadRows += 1
        self.assertTrue(len(writtenRows), numReadRows)

    def testReadWithCondition(self):
        testTable = self._database.getSameAsFieldTable()

        writtenRows = self._createSameAsFieldTableRows()
        testTable.write(writtenRows)

        parameterProvider = self.SameAsFieldTableParameterProvider()
        condition = "name='Name1'"
        readRows = testTable.read(parameterProvider, condition)
        expectedRowNum = 1
        for readRow in readRows:
            self._checkSameAsFieldTableRow(writtenRows[expectedRowNum], readRow)

    def testUpdate(self):
        testTable = self._database.getSameAsFieldTable()

        writtenRows = self._createSameAsFieldTableRows()
        testTable.write(writtenRows)

        updateRowId = 3
        updateRow = self._createSameAsFieldTableRow(updateRowId, "UpdatedName")
        updateCondition = "id=" + str(updateRowId)
        testTable.update(updateRow, updateCondition)

        parameterProvider = self.SameAsFieldTableParameterProvider()
        readRows = testTable.read(parameterProvider, updateCondition)
        for readRow in readRows:
            self._checkSameAsFieldTableRow(updateRow, readRow)

    def _createSameAsFieldTableRows(self):
        rows = []
        for rowId in range(NUM_SIMPLE_PARAM_TABLE_ROWS):
            rows.append(self._createSameAsFieldTableRow(rowId, "Name" + str(rowId)))

        return rows

    def _createSameAsFieldTableRow(self, rowId, name):
        values = [rowId for i in range(SAME_AS_FIELD_TABLE_COUNT)]
        testBlob = self.api.explicit_same_as_field.TestBlob.fromFields(len(values), values)

        valuesExplicit = [rowId + 1 for i in range(SAME_AS_FIELD_TABLE_COUNT_EXPLICIT)]
        testBlobExplicit = self.api.explicit_same_as_field.TestBlob.fromFields(len(valuesExplicit),
                                                                               valuesExplicit)

        return (rowId, name, SAME_AS_FIELD_TABLE_COUNT, testBlob, testBlobExplicit)

    def _checkSameAsFieldTableRow(self, row1, row2):
        self.assertEqual(row1, row2)

    def _isTableInDb(self):
        # check if database does contain table
        sqlQuery = "SELECT name FROM sqlite_master WHERE type='table' AND name='" + TABLE_NAME + "'"
        for row in self._database.connection().cursor().execute(sqlQuery):
            if len(row) == 1 and row[0] == TABLE_NAME:
                return True

        return False

TABLE_NAME = "sameAsFieldTable"
NUM_SIMPLE_PARAM_TABLE_ROWS = 5
SAME_AS_FIELD_TABLE_COUNT = 10
SAME_AS_FIELD_TABLE_COUNT_EXPLICIT = 11

File: test/language/explicit_parameters/cpp/ExplicitSameAsFieldTest.cpp
#include <cstdio>
#include <string>
#include <vector>

#include "explicit_parameters/ExplicitParametersDb.h"
#include "gtest/gtest.h"
#include "zserio/RebindAlloc.h"
#include "zserio/StringConvertUtil.h"

namespace explicit_parameters
{
namespace explicit_same_as_field
{

using allocator_type = ExplicitParametersDb::allocator_type;
using string_type = zserio::string<allocator_type>;
template <typename T>
using vector_type = zserio::vector<T, allocator_type>;

class ExplicitSameAsFieldTest : public ::testing::Test
{
public:
    ExplicitSameAsFieldTest()
    {
        std::remove(DB_FILE_NAME);

        m_database = new ExplicitParametersDb(DB_FILE_NAME);
        m_database->createSchema();
    }

    ~ExplicitSameAsFieldTest() override
    {
        delete m_database;
    }

    ExplicitSameAsFieldTest(const ExplicitSameAsFieldTest&) = delete;
    ExplicitSameAsFieldTest& operator=(const ExplicitSameAsFieldTest&) = delete;

    ExplicitSameAsFieldTest(ExplicitSameAsFieldTest&&) = delete;
    ExplicitSameAsFieldTest& operator=(ExplicitSameAsFieldTest&&) = delete;

protected:
    void fillSameAsFieldTableRow(SameAsFieldTable::Row& row, uint32_t id, const string_type& name)
    {
        row.setId(id);
        row.setName(name);
        row.setCount(SAME_AS_FIELD_TABLE_COUNT);

        TestBlob testBlob;
        vector_type<uint8_t>& values = testBlob.getValues();
        for (uint32_t i = 0; i < SAME_AS_FIELD_TABLE_COUNT; ++i)
        {
            values.push_back(static_cast<uint8_t>(id));
        }
        row.setBlob(testBlob);

        TestBlob testBlobExplicit;
        vector_type<uint8_t>& valuesExplicit = testBlobExplicit.getValues();
        for (uint32_t i = 0; i < SAME_AS_FIELD_TABLE_COUNT_EXPLICIT; ++i)
        {
            valuesExplicit.push_back(static_cast<uint8_t>(id + 1));
        }
        row.setBlobExplicit(testBlobExplicit);
    }

    void fillSameAsFieldTableRows(vector_type<SameAsFieldTable::Row>& rows)
    {
        rows.clear();
        rows.reserve(NUM_SAME_AS_FIELD_TABLE_ROWS);
        for (uint32_t id = 0; id < NUM_SAME_AS_FIELD_TABLE_ROWS; ++id)
        {
            const string_type name = "Name" + zserio::toString<allocator_type>(id);
            SameAsFieldTable::Row row;
            fillSameAsFieldTableRow(row, id, name);
            rows.push_back(row);
        }
    }

    static void checkSameAsFieldTableRow(const SameAsFieldTable::Row& row1, const SameAsFieldTable::Row& row2)
    {
        ASSERT_EQ(row1.getId(), row2.getId());
        ASSERT_EQ(row1.getName(), row2.getName());
        ASSERT_EQ(row1.getCount(), row2.getCount());
        ASSERT_EQ(row1.getBlob(), row2.getBlob());
        ASSERT_EQ(row1.getBlobExplicit(), row2.getBlobExplicit());
    }

    static void checkSameAsFieldTableRows(
            const vector_type<SameAsFieldTable::Row>& rows1, const vector_type<SameAsFieldTable::Row>& rows2)
    {
        ASSERT_EQ(rows1.size(), rows2.size());
        for (size_t i = 0; i < rows1.size(); ++i)
        {
            checkSameAsFieldTableRow(rows1[i], rows2[i]);
        }
    }

    class SameAsFieldTableParameterProvider : public SameAsFieldTable::IParameterProvider
    {
    public:
        uint32_t getCount(SameAsFieldTable::Row&) override
        {
            return SAME_AS_FIELD_TABLE_COUNT_EXPLICIT;
        }
    };

    ExplicitParametersDb* m_database;

    static const char* const DB_FILE_NAME;

    static const uint32_t NUM_SAME_AS_FIELD_TABLE_ROWS;
    static const uint32_t SAME_AS_FIELD_TABLE_COUNT;
    static const uint32_t SAME_AS_FIELD_TABLE_COUNT_EXPLICIT;
};

const char* const ExplicitSameAsFieldTest::DB_FILE_NAME =
        "language/explicit_parameters/explicit_same_as_field_test.sqlite";

const uint32_t ExplicitSameAsFieldTest::NUM_SAME_AS_FIELD_TABLE_ROWS = 5;
const uint32_t ExplicitSameAsFieldTest::SAME_AS_FIELD_TABLE_COUNT = 10;
const uint32_t ExplicitSameAsFieldTest::SAME_AS_FIELD_TABLE_COUNT_EXPLICIT = 11;

TEST_F(ExplicitSameAsFieldTest, readWithoutCondition)
{
    SameAsFieldTable& sameAsFieldTable = m_database->getSameAsFieldTable();

    SameAsFieldTableParameterProvider parameterProvider;
    vector_type<SameAsFieldTable::Row> writtenRows;
    fillSameAsFieldTableRows(writtenRows);
    sameAsFieldTable.write(parameterProvider, writtenRows);

    SameAsFieldTable::Reader reader = sameAsFieldTable.createReader(parameterProvider);

    vector_type<SameAsFieldTable::Row> readRows;
    while (reader.hasNext())
    {
        readRows.push_back(reader.next());
    }
    checkSameAsFieldTableRows(writtenRows, readRows);
}

TEST_F(ExplicitSameAsFieldTest, readWithCondition)
{
    SameAsFieldTable& sameAsFieldTable = m_database->getSameAsFieldTable();

    SameAsFieldTableParameterProvider parameterProvider;
    vector_type<SameAsFieldTable::Row> writtenRows;
    fillSameAsFieldTableRows(writtenRows);
    sameAsFieldTable.write(parameterProvider, writtenRows);

    const string_type condition = "name='Name1'";
    SameAsFieldTable::Reader reader = sameAsFieldTable.createReader(parameterProvider, condition);

    ASSERT_TRUE(reader.hasNext());
    SameAsFieldTable::Row readRow = reader.next();
    ASSERT_FALSE(reader.hasNext());

    const size_t expectedRowNum = 1;
    checkSameAsFieldTableRow(writtenRows[expectedRowNum], readRow);
}

TEST_F(ExplicitSameAsFieldTest, update)
{
    SameAsFieldTable& sameAsFieldTable = m_database->getSameAsFieldTable();

    SameAsFieldTableParameterProvider parameterProvider;
    vector_type<SameAsFieldTable::Row> writtenRows;
    fillSameAsFieldTableRows(writtenRows);
    sameAsFieldTable.write(parameterProvider, writtenRows);

    const uint64_t updateRowId = 3;
    SameAsFieldTable::Row updateRow;
    fillSameAsFieldTableRow(updateRow, updateRowId, "UpdatedName");
    const string_type updateCondition = "id=" + zserio::toString<allocator_type>(updateRowId);
    sameAsFieldTable.update(parameterProvider, updateRow, updateCondition);

    SameAsFieldTable::Reader reader = sameAsFieldTable.createReader(parameterProvider, updateCondition);
    ASSERT_TRUE(reader.hasNext());
    SameAsFieldTable::Row readRow = reader.next();
    ASSERT_FALSE(reader.hasNext());
    checkSameAsFieldTableRow(updateRow, readRow);
}

} // namespace explicit_same_as_field
} // namespace explicit_parameters
